server/storage/storage_test.go

```diff
@@ -279,3 +281,166 @@ func TestLoadRegionsExceedRangeLimit(t *testing.T) {
 	}
 	re.NoError(failpoint.Disable("github.com/tikv/pd/server/storage/kv/withRangeLimit"))
 }
+
+const (
+	keyChars = "abcdefghijklmnopqrstuvwxyz"
+	keyLen   = 20
+)
+
+func generateKeys(size int) []string {
+	m := make(map[string]struct{}, size)
+	for len(m) < size {
+		k := make([]byte, keyLen)
+		for i := range k {
+			k[i] = keyChars[rand.Intn(len(keyChars))]
+		}
+		m[string(k)] = struct{}{}
+	}
+
+	v := make([]string, 0, size)
+	for k := range m {
+		v = append(v, k)
+	}
+	sort.Strings(v)
+	return v
+}
+
+func randomMerge(regions []*metapb.Region, n int, ratio int) {
+	rand.Seed(6)
+	note := make(map[int]bool)
+	for i := 0; i < n*ratio/100; i++ {
+		pos := rand.Intn(n)
+		for {
+			if _, ok := note[pos]; !ok {
+				break
+			}
+			pos = rand.Intn(n)
+		}
+		note[pos] = true
+
+		mergeIndex := pos
+		if mergeIndex == 0 {
+			for {
+				mergeIndex++
+				if _, ok := note[mergeIndex]; !ok {
+					break
+				}
+			}
+			regions[mergeIndex].StartKey = regions[pos].StartKey
+		} else {
+			for {
+				mergeIndex--
+				if _, ok := note[mergeIndex]; !ok {
+					break
+				}
+			}
+			if mergeIndex != -1 {
+				regions[pos].StartKey = regions[mergeIndex].StartKey
```

Comment:
```suggestion
				regions[pos].EndKey = regions[mergeIndex].EndKey
```
It looks like it should be the end key here?

Reply:
When we find a "mergeIndex" before "pos", I think we should reset the pos region StartKey to the mergeIndex region StartKey for the purpose of merging mergeIndex region to pos region.
```
Origin:

Region(mergeIndex)                             Region(pos)
|-------------------------------------------|  |--------------------------------------------|
mergeIndex(StartKey)       mergeIndex(EndKey)  pos(StartKey)                      pos(EndKey)


Reset:

Region(mergeIndex)
|-------------------------------------------| 
mergeIndex(StartKey)       mergeIndex(EndKey)
Region(pos)
|------------------------------------------------------------------------------------------|
mergeIndex(StartKey)                                                             pos(EndKey)
```

And when I set "egions[pos].EndKey = regions[mergeIndex].EndKey", the benchmark log is (input is 100):
```
goos: darwin
goarch: amd64
pkg: github.com/tikv/pd/server/storage
cpu: Intel(R) Core(TM) i5-6267U CPU @ 2.90GHz
BenchmarkLoadRegionsByRandomMerge/merge_ratio_0-4         	1000000000	         0.001031 ns/op	       0 B/op	       0 allocs/op
--- BENCH: BenchmarkLoadRegionsByRandomMerge/merge_ratio_0-4
    storage_test.go:408: region number is:  100
    ...
BenchmarkLoadRegionsByRandomMerge/merge_ratio_20-4        	1000000000	         0.0007170 ns/op	       0 B/op	       0 allocs/op
--- BENCH: BenchmarkLoadRegionsByRandomMerge/merge_ratio_20-4
    storage_test.go:408: region number is:  100
    ...
BenchmarkLoadRegionsByRandomMerge/merge_ratio_40-4        	1000000000	         0.0009663 ns/op	       0 B/op	       0 allocs/op
--- BENCH: BenchmarkLoadRegionsByRandomMerge/merge_ratio_40-4
    storage_test.go:408: region number is:  100
    ...
BenchmarkLoadRegionsByRandomMerge/merge_ratio_60-4        	1000000000	         0.001285 ns/op	       0 B/op	       0 allocs/op
--- BENCH: BenchmarkLoadRegionsByRandomMerge/merge_ratio_60-4
    storage_test.go:408: region number is:  96
    ...
BenchmarkLoadRegionsByRandomMerge/merge_ratio_80-4        	1000000000	         0.0007771 ns/op	       0 B/op	       0 allocs/op
--- BENCH: BenchmarkLoadRegionsByRandomMerge/merge_ratio_80-4
    storage_test.go:408: region number is:  96
    ...
PASS
```
The region number is not correct. For example, when ration is 20, the region number should be 80